packages/pds/src/migrate-script/index.ts

```diff
@@ -0,0 +1,417 @@
+}
+
+const doImport = async (
+  ctx: AppContext,
+  db: MigrateDb,
+  pds: PdsInfo,
+  did: string,
+  since?: string,
+) => {
+  const storage = new SqlRepoStorage(ctx.db, did)
+  const root = await storage.getRootDetailed()
+  if (!root) {
+    throw new Error(`repo not found: ${did}`)
+  }
+  if (since && root.rev === since) {
+    return
+  }
+  const carStream = await storage.getCarStream(since)
+
+  const importRes = await axios.post(
+    `${pds.url}/xrpc/com.atproto.temp.importRepo`,
+    carStream,
+    {
+      params: { did },
+      headers: { 'content-type': 'application/vnd.ipld.car' },
+      decompress: true,
+      responseType: 'stream',
+    },
+  )
+
+  for await (const log of importRes.data) {
+    if (typeof log === 'string') {
+      const lines = log.split('\n')
+      for (const line of lines) {
+        if (line.indexOf('failed to import blob') > 0) {
+          const cid = line.split(':')[1].trim()
+          await logFailedBlob(db, did, cid)
+        }
+      }
+    }
+    console.log(`import update for ${did}: `, log.toString())
+  }
+  return root.rev
+}
+
+const lockAndTransfer = async (
+  ctx: AppContext,
+  db: MigrateDb,
+  pds: PdsInfo,
+  status: Status,
+) => {
+  const defer = createDeferrable()
+  ctx.db
+    .transaction(async (dbTxn) => {
+      const storage = new SqlRepoStorage(dbTxn, status.did)
+      await storage.lockRepo()
+      await defer.complete
+    })
+    .catch((err) => {
+      console.error(`error in repo lock tx for did: ${status.did}`, err)
+    })
+
+  let importedRev
+  try {
+    importedRev = await doImport(
+      ctx,
+      db,
+      pds,
+      status.did,
+      status.importedRev ?? undefined,
+    )
+
+    const lastOp = await ctx.plcClient.getLastOp(status.did)
+    if (!lastOp || lastOp.type === 'plc_tombstone') {
+      throw new Error('could not find last plc op')
+    }
+    const plcOp = await plcLib.createUpdateOp(
+      lastOp,
+      ctx.plcRotationKey,
+      (normalized) => {
+        if (!status.signingKey) {
+          throw new Error('no reserved signing key')
+        }
+        return {
+          ...normalized,
+          verificationMethods: {
+            atproto: status.signingKey,
+          },
+          services: {
+            atproto_pds: {
+              type: 'AtprotoPersonalDataServer',
+              endpoint: pds.url,
+            },
+          },
+        }
+      },
+    )
+
+    const accountRes = await getUserAccount(ctx, status.did)
+    await axios.post(`${pds.url}/xrpc/com.atproto.temp.transferAccount`, {
+      did: status.did,
+      handle: accountRes.handle,
+      plcOp,
+    })
+
+    defer.resolve()
+
+    await ctx.db.db
+      .updateTable('user_account')
+      .where('did', '=', status.did)
+      .set({ pdsId: pds.id })
+      .execute()
+    await ctx.db.db
+      .updateTable('repo_root')
+      .where('did', '=', status.did)
+      .set({ did: `migrated-${status.did}` })
+      .execute()
```

Comment:
These are unlikely to fail, but it's tempting to wrap these in a transaction.

Reply:
yeah good call - also attempted to track the state where these fail but the rest of the transfer succeeds since that'd be a pretty subtle failure case